### PR TITLE
raftstore-v2: cleanup stale tablet on restart

### PR DESCRIPTION
**Pull request by BusyJay**
### What is changed and how it works?
Issue Number: Ref #12842 

What's Changed:

```commit-message
If operations like snapshot, split, are aborted by restart, they needs
to be either resumed or cleanup. This PR checks for garbage after
restart and resume committed operations.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore-v2/src/operation/command/admin/split.rs [hunk 1]
```diff
@@ -343,16 +349,19 @@ impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
...
-            .create_at(&derived_path, None, 0)
-            .unwrap_or_else(|e| {
-                panic!(
-                    "{:?} fails to create checkpoint with path {:?}: {:?}",
-                    self.logger.list(),
-                    derived_path,
-                    e
-                )
-            });
+        // If it's recovered from restart, it's possible the target path exists already.
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/command/admin/split.rs:352]: Add a comment somewhere to explain checkpoint is atomic, if there isn't one already.

File: components/raftstore-v2/src/operation/ready/snapshot.rs [hunk 2]
```diff
@@ -115,6 +116,48 @@ impl Debug for GenSnapTask {
...
+pub fn install_tablet<EK>(
+    registry: &TabletRegistry<EK>,
+    source: &Path,
+    region_id: u64,
+    tablet_index: u64,
+) -> bool {
+    let target_path = registry.tablet_path(region_id, tablet_index);
+    if !source.exists() {
+        return false;
+    }
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/ready/snapshot.rs:142]:
```suggestion
    if !source.exists() {
        return false;
    }
    let target_path = registry.tablet_path(region_id, tablet_index);
```

**Review by tabokie — APPROVED** [review]

**Review by ethercflow — APPROVED** [review]
LGTM

File: components/raftstore-v2/src/operation/command/mod.rs [hunk 3]
```diff
@@ -505,10 +507,14 @@ impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
 
     #[inline]
     pub fn flush(&mut self) {
+        let (index, term) = self.apply_progress();
+        let flush_state = self.flush_state().clone();
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/command/mod.rs:511]: Why clone?

File: components/raftstore-v2/src/raft/apply.rs [hunk 4]
```diff
@@ -115,20 +117,21 @@ impl<EK: KvEngine, R> Apply<EK, R> {
...
-        self.flush_state.set_applied_index(index);
+        self.applied_index = index;
         self.applied_term = term;
         if self.log_recovery.is_none() {
             return;
         }
         let log_recovery = self.log_recovery.as_ref().unwrap();
         if log_recovery.iter().all(|v| index >= *v) {
             self.log_recovery.take();
+            let _ = self.tablet.flush_cfs(&[], false);
```
Comment (tabokie) [inline components/raftstore-v2/src/raft/apply.rs:128]: Needs some comments, I don't know why.

File: components/raftstore-v2/src/fsm/peer.rs [hunk 5]
```diff
@@ -194,6 +195,13 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: Transport> PeerFsmDelegate<'a, EK, ER,
         if self.fsm.peer.storage().is_initialized() {
             self.fsm.peer.schedule_apply_fsm(self.store_ctx);
         }
+        // Unlike v1, it's a must to set ready when there are pending entries. Otherwise
+        // it may block for ever when there is unapplied conf change.
+        let entry_storage = self.fsm.peer.storage().entry_storage();
+        if entry_storage.commit_index() > entry_storage.applied_index() || self.fsm.peer.is_leader()
```
Comment (tabokie) [inline components/raftstore-v2/src/fsm/peer.rs:201]: `is_leader` case is not explained in the comment.

**Review by tabokie — APPROVED** [review]

File: components/raftstore-v2/src/operation/command/mod.rs (hunk 3 above)
Comment (BusyJay) [reply components/raftstore-v2/src/operation/command/mod.rs:511]: Because writebatch is mut borrow next line.

**Pull request merged**